Add AutoPackets documentation and fix navigation issue hiding Multiplayer Gem docs (#849)
o3de/o3de.org: content/docs/user-guide/gems/reference/multiplayer/multiplayer-gem/autopackets.md

---

> title: Multiplayer AutoPackets

Reviewer: Similar to auto-components I think AutoPacket = Auto-Packet or Auto-packet?

description: A reference for defining Open 3D Engine multiplayer packets through autopackets.
linktitle: AutoPackets
---

> *AutoPackets* provide a convenient way to define network packets used to communicate data between endpoints in a Multiplayer session. Using the [AzAutoGen](/docs/user-guide/engine/autogen) system, autopacket files found inside of your project are processed during builds to create C++ classes and handlers for packets with defined payloads.

Reviewer 2: To me its not clear what an auto-packet is here? Are we defining a protocol for communication? Are they are collecting of attributes for a packet?

Could you provide an example(s) of a 'typical' autopacket here ie a Connect packet that contains the properties expected during connection events such as a protocol version and ticket id?

Author: There's an example detailing that further down in the doc.

In order to enable autopacket builds for your project, follow the instructions in [Multiplayer Project Configuration](./configuration).

> ## AutoPacket file structure

Author: Not to my knowledge, unless there was one for AutoComponents that's not apparent to me?

> AutoPackets are defined in XML files, placed in the `Code\Source\Autogen` directory of the Multiplayer Gem.

Author: They can be anywhere making use of AzNetworking including AzNetworking itself, I'll do some restructuring here.

### PacketGroup attributes

The PacketGroup defines a set of packets that are identified in increasing numerical order

| Property | Description | Type |
|---|---|---|---|
| Name | The class name of the group of generated autopacket. | `string`: Must be a valid C++ class name. |
| PacketStart | The value from which to begin IDing packets, typically CorePackets::PacketType::Max. | `PacketType`: Must be a PacketType value. |

### Packet attributes

> The `Packet` tag defines the name, description and handshake relevance of the packet
> 
> | Property | Description | Type |
> |---|---|---|---|
> | Name | The class name of the generated autopacket. | `string`: Must be a valid C++ class name. |
> | Desc | The description of the generated autopacket. | `string`: A user friendly string description of the packet. |
> | HandshakePacket | If the packet is part of a connection handshake. | `string`: Only packets with HandshakePacket as true can be processed while the user implemented IsHandshakeComplete function returns false. |

Reviewer 3: ```suggestion
The `Packet` tag defines a new packet.

| Property | Description | Type |
|---|---|---|---|
| Name | The class name of the generated autopacket. | A valid C++ class identifier. |
| Desc | The description of the generated autopacket. | `string` |
| HandshakePacket | If the packet is part of a connection handshake. | `bool` |

{{< note >}}
Packets with `HandshakePacket ` set as true will only be processed if the user-implemented `IsHandshakeComplete` function for the packet returns `false`.
{{< /note >}}
```

### Member attributes

> The `Member` tag defines a member on the packet. This is the primary mechanism for defining a packet's payload.
> 
> | Property | Description | Type |
> |---|---|---|---|
> | Name | The name of the packet member. | `string`: Must be a valid C++ variable name. |
> | Type | The type of the packet member. | `string`: Must be a valid C++ type. |
> | Init | The initial value of the packet member. | `<Type>`: Must be a valid value of `<Type>`. |
> | Container | If specified, the member represents a container of `<Type>`. | `Vector`, `Array`|
> | | **Vector**: A resizeable vector. |
> | | **Array**: A fixed size array. | |
> | Count | If Container is specified, the size of the container. | `uint` |

Reviewer 3: ```suggestion
The `Member` tag defines data on the packet. This is the primary mechanism for defining a packet's payload.

| Property | Description | Type |
|---|---|---|---|
| Name | The name of the packet member. | A valid C++ variable name. |
| Type | The type of the packet member. | A valid C++ type. |
| Init | The initial value of the packet member. | `<Type>` |
| Container | If provided, the member is a container of `<Type>`. | `Vector`, `Array`|
| | **Vector**: A resizeable vector. |
| | **Array**: A fixed size array. | |
| Count | If Container is provided, the size of the container. | `uint` |
```
